src/snapshot/resource_map.rs

```diff
@@ -0,0 +1,85 @@
+use std::marker::PhantomData;
+
+use bevy::{
+    ecs::entity::{EntityMapper, MapEntities},
+    prelude::*,
+};
+
+use crate::{LoadWorld, LoadWorldSet, RollbackEntityMap};
+
+/// A [`Plugin`] which updates the state of a post-rollback [`Resource`] `R` using [`MapEntities`].
+pub struct GgrsResourceMapEntitiesPlugin<R>
+where
+    R: Resource + MapEntities,
+{
+    _phantom: PhantomData<R>,
+}
+
+impl<R> Default for GgrsResourceMapEntitiesPlugin<R>
+where
+    R: Resource + MapEntities,
+{
+    fn default() -> Self {
+        Self {
+            _phantom: default(),
+        }
+    }
+}
+
+impl<R> GgrsResourceMapEntitiesPlugin<R>
+where
+    R: Resource + MapEntities,
+{
+    /// Exclusive system which will apply a [`RollbackEntityMap`] to the [`Resource`] `R`, provided it implements [`MapEntities`].
+    pub fn update(world: &mut World) {
+        world.resource_scope(|world: &mut World, map: Mut<RollbackEntityMap>| {
+            apply_rollback_map_to_resource_inner::<R>(world, map);
+        });
+    }
+}
+
+fn apply_rollback_map_to_resource_inner<R>(world: &mut World, map: Mut<RollbackEntityMap>)
+where
+    R: Resource + MapEntities,
+{
+    let mut applied_entity_map = map.get_map();
```

Comment:
The problem is every mapping operation requires a mutable reference to the entity map (to allow for creating dead entities). A way we could mitigate this problem would be to actually create the `RollbackHierarchy` plugin I mentioned previously, where parent-child rollback is done through commands rather than entity mapping and reflection.

Honestly, I think the change to entity mapping Bevy made in 0.11 made the feature quite hostile for every case that isn't explicitly moving from one world to another. Instead of providing a mutable reference to an entity map, they probably should've passed it a `Commands` to allow the mapper to create entities as required explicitly.

Reply:
I see... Maybe EntityMapper should have been a trait instead?

Reply:
Agreed. For now, I think we should leave the cloning in, and once this is merged I'll see if I can get a PR merged into Bevy which gives us some kind of alternative to mutable `EntityMapper` which suits are use-case better.

Reply:
Or, we remove the cloning but all mapping operations need to happen sequentially. Personally, I think cloning would be more performant and ergonomic to use, but that's more of a hunch than anything else.